archunit/src/jdk15test/java/com/tngtech/archunit/core/importer/testexamples/methodresolution/DeterminesMethodAnalogouslyToReflectionApi.java

```diff
@@ -0,0 +1,874 @@
+                child.target();
+            }
+        }
+
+        public static class RightLeftOnGrandparentHasPrecedenceOverLeftAndRightOnParent {
+            interface LeftLeftGrandParent {
+            }
+
+            interface LeftRightGrandParent {
+            }
+
+            interface LeftParent {
+                void target();
+            }
+
+            interface RightLeftGrandParent {
+                @ExpectedMethod
+                void target();
+            }
+
+            interface RightRightGrandParent {
+                void target();
+            }
+
+            interface RightParent {
+                void target();
+            }
+
+            abstract static class GrandParentClass implements LeftLeftGrandParent, LeftRightGrandParent, RightLeftGrandParent, RightRightGrandParent {
+            }
+
+            abstract static class ParentClass extends GrandParentClass implements LeftParent, RightParent {
+            }
+
+            abstract static class Child extends ParentClass {
+            }
+
+            void scenario() {
+                Child child = null;
+                child.target();
+            }
+        }
+
+        public static class RightRightOnGrandparentHasPrecedenceOverLeftAndRightOnParent {
+            interface LeftLeftGrandParent {
+            }
+
+            interface LeftRightGrandParent {
+            }
+
+            interface LeftParent {
+                void target();
+            }
+
+            interface RightLeftGrandParent {
+            }
+
+            interface RightRightGrandParent {
+                @ExpectedMethod
+                void target();
+            }
+
+            interface RightParent {
+                void target();
+            }
+
+            abstract static class GrandParentClass implements LeftLeftGrandParent, LeftRightGrandParent, RightLeftGrandParent, RightRightGrandParent {
+            }
+
+            abstract static class ParentClass extends GrandParentClass implements LeftParent, RightParent {
+            }
+
+            abstract static class Child extends ParentClass {
+            }
+
+            void scenario() {
+                Child child = null;
+                child.target();
+            }
+        }
+    }
+
+    public static class MoreSpecificReturnTypeHasPrecedence {
+
+        public static class RightWithMoreSpecificReturnOnParentTypeHasPrecedenceOverAllOthers {
+            interface LeftLeftGrandParent {
+                Serializable target();
+            }
+
+            interface LeftRightGrandParent {
+                Serializable target();
+            }
+
+            interface LeftParent {
+            }
+
+            interface RightLeftGrandParent {
+                Serializable target();
+            }
+
+            interface RightRightGrandParent {
+                Serializable target();
+            }
+
+            interface RightParent {
+                @ExpectedMethod
+                String target();
```

Comment:
The way you implemented the resolve process it makes sense that this is the expected method, IF we know the `child.target()` call returns a String, which I presume we know from the byte code. Is it on byte code level the same that the most specific return type is chosen?

Reply:
In the bytecode of `Child` you won't find any trace of `target()`. Because `Child` is abstract and doesn't override `target()`. However, the compiler will always forbid you to override a method with a less specific return type. So whoever will eventually override / implement this method, the declaration will have the more specific return type. From that point of view it makes sense to pick the method with the most specific return type in the hierarchy (and again it's also what the Reflection API does, so no uncertainty there :man_shrugging: )

Reply:
Yes you are right, `Child` does not know about it. But the call of `scenario()` seems to know that `child.target()` is of return type String (I checked that in the byte code). Does this mean the compiler also resolves the more specific type?

Reply:
Yes, otherwise covariant return type overrides wouldn't work I guess, right? Once it's covariantly overridden in the hierarchy the more specific return type is the return type of the method so it can be assigned/used safely wherever the more specific type is needed (under the hood a lot more uglyness is going on, like bridge-methods :joy:)